### DISCUSSION TITLE
Discussion regarding the editing of the Yao website

### DISCUSSION THREAD

**Discussion by VarLad** [QuantumBFS/Yao.jl#282] [Ideas]
@Roger-luo 
Lets discuss a few ideas we can use to better edit the Yao website (the tutorials part)

**Comment by VarLad** [comment]
The simplest way I could think of, was to have a separate repo for tutorials (we can just use https://github.com/QuantumBFS/tutorials)

There we can just fork and edit the MITs191 github repo and deploy it to `quantumbfs.github.io/tutorials`

In the header of yaoquantum.org, we can redirect the `tutorials` to our above created page

**Reply by VarLad** [reply]
That way we can better/more easily customize the tutorials part. The remainder of the Yao site doesn't have any problems

**Reply by VarLad** [reply]
Franklin's support for Pluto notebooks would be coming soon
We can do this until then

**Reply by Roger-luo** [reply]
we were using a seperate webpage for tutorial before, the problem is how you make the banner consistent, the banner implementation was copy-pasted over a few repos for this reason, so if we want to host things in seperate repo, we need to think about how to combine them together. 

For the sake of consistency in style and rendering and CI, I think the notebook scripts should be stay in tutorial repo, and in the main website CI, it will clone this repo and build the HTML from it.

**Reply by Roger-luo** [reply]
> There we can just fork and edit the MITs191 github repo and deploy it to quantumbfs.github.io/tutorials

this doesn't really work well - I've done the fork, it looks strange when two different CSS coexists

**Reply by VarLad** [reply]
@Roger-luo One solution I can think of, is to keep header persistent throughout the tutorials

As in, it won't disappear as you keep scrolling down, in tutorials
Another thing, I'd like to do, is to cut out out a bit of the footer for the tutorial page, as if I make that persistent, it'd take up quite a bit of space
We can just include the  the slack link and the issues link in the header
All of this, specifically for the tutorials page. No change to the remainder of the site whatsoever

We can keep the below information either in the welcome page, or somewhere in the side bar
```
The Yao Framework

Extensible Efficient Quantum Algorithm Design For Humans.
© QuantumBFS. Website built with Franklin.jl and the Julia programming language.
```

**Reply by Roger-luo** [reply]
I think a good reference is pennylane's website: https://pennylane.readthedocs.io/en/stable/

It's written by professional frontend developers, so we could use this as a reference.

**Reply by VarLad** [reply]
How about https://qiskit.org/textbook/ch-states/case-for-quantum.html
？

**Reply by Roger-luo** [reply]
Yeah this one looks great too, I guess the problem is our banner has a very inconsistent style with the sidebar... and we don't have a good UI on the banner to guide people to other places comparing to this IBM page.

**Reply by Roger-luo** [reply]
would be nice if you could implement a small demo page @VarLad I think you could post the link of the demo page here. 

> One solution I can think of, is to keep header persistent throughout the tutorials

Yeah, I agree, this would help people go to other pages.

> We can keep the below information either in the welcome page, or somewhere in the side bar

True. I think we can have a smaller footer for the tutorial page and other content dense page.

**Comment by VarLad** [comment]
@Roger-luo Take a look at this sample https://varlad.github.io/Spring21/x/
The repo is https://github.com/VarLad/varlad.github.io

This implementation is stable, with the banner and everything. I don't think we need the footer（We can easily include that information in the sidebar and that saves space）

The only problem I'm facing is padding (a part of sidebar and notebook get covered by the header)

Since we're using custom css and a different template here, I think it'd be easier to deploy the tutorials on a different repo. That way, anyone can customize it _without_ caring about how it affects the site

Any ideas regarding the two problems are welcome

**Reply by VarLad** [reply]
Please confirm if you agree with this approach
If you agree, all thats left is to tweak the padding and to fill out the incomplete contents

**Reply by VarLad** [reply]
Unitary Hack will be over soon. I hope I can complete this in 24 hours

**Reply by Roger-luo** [reply]
I think the current tutorial page looks good in general, but need to make the sidebar style consistent with banner (not use the MIT course CSS)

![image](https://user-images.githubusercontent.com/8445510/120054250-db7d0700-bffc-11eb-9467-40ad631e4deb.png)

> Since we're using custom css and a different template here, I think it'd be easier to deploy the tutorials on a different repo. That way, anyone can customize it without caring about how it affects the site

the other problem is we need to setup the pipeline for the tutorial where we:

1. build the tutorial page along with homepage in one repo
2. host the tutorial pluto script files in the tutorial repo so that one can easily download and run locally

> Unitary Hack will be over soon. I hope I can complete this in 24 hours

I don't think we can finish this in 24hrs, but if you have time in the summer, maybe consider apply our summer of code (but it will require you put more time into this tho)

**Comment by VarLad** [comment]
> the other problem is we need to setup the pipeline for the tutorial

@Roger-luo Take a look here 
https://github.com/VarLad/tutorials
If you push on my tutorials fork, that automatically pushes here: https://github.com/varlad/varlad.github.io
The contents of the `notebooks` folder of tutorials gets copied into the `notebooks` folder of the site

I consider that one complete. Is it okay if I make a PR to the tutorials repo?

**Reply by VarLad** [reply]
Also, I'd say its worth it to make some changes to use PlutoSliderServer, instead of PlutoBinder, in quantumbfs.github.io

**Reply by VarLad** [reply]
> I don't think we can finish this in 24hrs, but if you have time in the summer, maybe consider apply our summer of code (but it will require you put more time into this tho)

I'd rather prefer a more scientific/mathematical work in the summers though. :sweat_smile:

I'll try completing this by Unitary Hack 🙂. Of course I don't mind continuing after UnitaryHack ends too, but I guess I get the bounty if I complete the work before it ends
If not by 24 hours, then definitely by 48-72 hours (considering making the site more _consistent_ is the only work left)

**Reply by VarLad** [reply]
> I consider that one complete. Is it okay if I make a PR to the tutorials repo?

@Roger-luo Is that one good to make a PR？

**Reply by VarLad** [reply]
Specifically this https://github.com/VarLad/tutorials/blob/master/.github/workflows/push_to_site.yml

**Reply by Roger-luo** [reply]
Yes please always feel free to open a PR first I'll comment over there

**Reply by Roger-luo** [reply]
> I'd rather prefer a more scientific/mathematical work in the summers though. 😅

FYI. There are other projects, e.g https://github.com/QuantumBFS/SummerOfCode/blob/main/ideas/simulation.md

I haven't really advertise this because it came out in a rush and I won't be the main mentor because I won't have much time mentoring students this year. @ChenZhao44 will be the mentor instead.

**Reply by Roger-luo** [reply]
I'm OK grant the bounty if you are late than their deadline as long as you keep working on this, same for @Sov-trotter but since it won't be me paying you (but unitary fund, who needs to organize much more PRs), it would be better if you both try to finish it before the deadline.

**Reply by VarLad** [reply]
> FYI. There are other projects, e.g https://github.com/QuantumBFS/SummerOfCode/blob/main/ideas/simulation.md
 
Thanks for the information. But doesn't go with my current interests :sweat_smile: 
For my summer plan, I'll try looking into what are the things that can be done with RC using NiLang (That involves bugging GiggleLiu with some potential ideas I'd have)
Also, if there's a GSoC for more/better tutorials in QC/QML, that'd be nice (I'm currently making them anyway though :sweat_smile:)

**Reply by Sov-trotter** [reply]
Yeah @Roger-luo  I am happy to keep working on the few remaining issues.

**Reply by VarLad** [reply]
@Roger-luo I believe the current assigned work is over

**Reply by Roger-luo** [reply]
> I believe the current assigned work is over

IMHO It's far from finishing. The original unitaryhack issue has much more requirements.

**Reply by VarLad** [reply]
Oh
I thought making the site more consistent was the only work

```
    we need a more consistent solution to add banners to pages, see also (Documenter/#1132)
    improve large screen experience (Documenter/#1563)
    improve CSS to have a more consistent style.
```

and

```


ideally, we should move all documentation into one repo, and build them together, like DiffEq documentation, this requires

    a CI that gets triggered (or a nightly cron job) for the master branch to rebuild new documentation parts for each component packages
    integration with the front page
```
I guess

The Documentation isn't managed by Franklin, is it?

**Reply by Roger-luo** [reply]
> I thought making the site more consistent was the only work

but it's far from finishing right? the two Documenter issue is still open and you didn't improve the CSS at all. I can only say the CI pipeline is half done since the idea of triggering new tutorials builds kinda works, but the documentation and webpage are still separated.

**Reply by Roger-luo** [reply]
> The Documentation isn't managed by Franklin, is it?

no, that's part of the reason why this issue worth 100$

**Reply by VarLad** [reply]
Okay! Then I agree a lot of work is still left
Lets discuss the output a bit
@Roger-luo 

1) Improve CSS -> what would be the desired output? I think shrinking the sidebar is one. Any other ideas?

2) Documenter -> https://github.com/JuliaDocs/Documenter.jl/issues/1563 Do you just want to stretch it 

![Screenshot from 2021-06-01 00-16-13](https://user-images.githubusercontent.com/51269425/120231312-bee80580-c26e-11eb-8640-c5e17dedf2c8.png)

3) Finally, you want the navbar from yaoquantum.org in docs, and then to use one repo to deploy them?

**Reply by Roger-luo** [reply]
> Documenter -> JuliaDocs/Documenter.jl#1563 Do you just want to stretch it

I think it's also related to https://github.com/JuliaDocs/Documenter.jl/issues/1168, would be nice to ask there too since this will require you to submit a PR to Documenter, I'd like to have things centralized like https://pydantic-docs.helpmanual.io/

ideally, it should show more verbose TOC on large screen.

> Improve CSS -> what would be the desired output? I think shrinking the sidebar is one. Any other ideas?

also need to change the tutorial sidebar color and fonts etc. since it was copied from MIT course page. And for the documentation CSS, we need to come up with a similar color scheme with the main landing page, the default Documenter color scheme is quite far from the landing page color scheme. 

It would be nice to re-use some of the fonts and settings from https://github.com/JuliaDiff/DocThemeIndigo.jl

**Reply by Roger-luo** [reply]
> Finally, you want the navbar from yaoquantum.org in docs, and then to use one repo to deploy them?

yes, ideally the web contents of website, the documentation, the tutorial should be all in one repo, and gets triggered when the main Yao repo and tutorial notebook repo gets new commits in their master/main branch.

**Reply by Roger-luo** [reply]
the idea is to remove all the copy pasted banner HTML strings in `docs/make.jl`

**Reply by VarLad** [reply]
That would remove the navbar I guess. You don't need the navbar then, in the docs?

**Reply by Roger-luo** [reply]
> That would remove the navbar I guess. You don't need the navbar then, in the docs?

no, as I mentioned in the issue, you will need to implement https://github.com/JuliaDocs/Documenter.jl/issues/1132

the current implementation is only a workaround, that's why the code is copy pasted, instead of replacing the Documenter functions by writing monkey patch, we should do it by actually implementing it.

**Reply by VarLad** [reply]
The docs folder (https://github.com/QuantumBFS/Yao.jl/tree/master/docs), do you need a pipeline from it to `QuantumBFS/Quantumbfs.github.io/docs` folder or something, and then run make.jl from quantumbfs.github.io?

Or do you want to run the make.jl in Yao.jl, and deploy it to quantumbfs.github.io repo?

**Reply by VarLad** [reply]
@Roger-luo I believe we can achieve the header with Franklin, without having to implement it in Documenter


![Screenshot from 2021-06-01 04-18-01](https://user-images.githubusercontent.com/51269425/120247267-83f6c980-c290-11eb-9a67-aacd04e3337a.png)

**Reply by Roger-luo** [reply]
Are you doing it via iframe? If so, I don't think it's a good idea using iframe since it can easily break the style.

**Reply by VarLad** [reply]
I'm aware of that
But using with iframe-resizer, I've been able to get a somewhat consistent solution.  

I've figured out how to build the docs from the site and deploy them with the site but the solution would take some time to implement. I'll put the demo here when its done.

**Comment by VarLad** [comment]
Making a to-do list here

- [x] Making the tutorials render properly on a smartphone/smaller screen. MIT site does it by making the sidebar go on top. An alternative could be to build a toggler to hide it.
- [x] Using the `PlutoSliderServer` to create the htmls instead of the current PlutoBinder setup. 
- [x] Update everything to the latest. Latest version of Pluto provides the "Edit or run on binder" by default on the top
- [x] Fix some notebooks. Use the `PackageSpec` from Pkg
- [ ] Make the docs more compatible with the Yao site and banner.

**Reply by VarLad** [reply]
@Roger-luo I can start on the CI to integrate the Franklin header and deploying the docs with Franklin, after you're done merging the Quantumbfs.github.io and Yao.jl PRs

What do you say, will the task be considered complete, after thats done？🙂

**Reply by Roger-luo** [reply]
I think you still need to finish the two Documenter issues, so maybe you can work on that first, it is useful for broader Julia community

**Reply by VarLad** [reply]
> we need a more consistent solution to add banners to pages, see also (Documenter/#1132)
> improve large screen experience (Documenter/#1563)

I believe that integrating it with Franklin is easier and better. That way, the header and the docs will both render independently. As a result, when you make a change in the site's header, the docs won't be affected.

For the 2nd issue, I find that to be a bit more abstract

**Reply by Roger-luo** [reply]
> For the 2nd issue, I find that to be a bit more abstract

You will need to submit PR to Documenter with a proposal and let other developers there comment. 

> I believe that integrating it with Franklin is easier and better. That way, the header and the docs will both render independently. As a result, when you make a change in the site's header, the docs won't be affected.

The main issue is the use of `iframe` which I hope to get rid of.

**Reply by VarLad** [reply]
> The main issue is the use of iframe which I hope to get rid of.

Normally, its a pretty nice solution! The problems start when we start using them in containers with other other elements.
Also, that allows the the header to be independent of the docs. So when you change the Franklin header in the future, it'll just naturally work out!

![Screenshot from 2021-06-05 01-58-37](https://user-images.githubusercontent.com/51269425/120859444-d7735b00-c5a1-11eb-9a0e-0fcfe9a6dfe1.png)

**Reply by VarLad** [reply]
> You will need to submit PR to Documenter with a proposal and let other developers there comment.

Any particular proposal in mind?

**Reply by Roger-luo** [reply]
I think I have commented somewhere else, just add a more detailed table of content on the right hand side (which is entirely blank) and make the entire doc page centralized

**Reply by Roger-luo** [reply]
you don't even need to change the color and layout scheme, but just the general page layout

**Reply by Roger-luo** [reply]
It seems iframe may cause search issue for Google too: https://www.ostraining.com/blog/webdesign/against-using-iframes/

I think in general, the use of iframe is not recommended in best practice of frontends.

**Reply by VarLad** [reply]
Hmmmm.... I see. Regarding security issues, Mozilla pushed forward a solution to isolate them.

Also, my point being, my iframe setup is too simple to cause problems (for the Documenter part). As for the tutorials part, iframes is kinda broken there. We need a better iframes solution there or if possible, an iframes free solution (which I think is possible, but would require a lot more experimentation)

**Reply by VarLad** [reply]
> just add a more detailed table of content on the right hand side (which is entirely blank) and make the entire doc page centralized

Okay, so display a list of all the methods in the page in the right side?

**Reply by Roger-luo** [reply]
I think to have an iframe free solution, we just need to revise Pluto export function to generate the head and `div` seperately, but I'm not sure how.

For documenter I think asking Documenter to export markdown and let franklin render it would also be a solution.

**Reply by Roger-luo** [reply]
by table of content I mean the markdown table of content, documenter only display upto 2-level headers, and by more detailed TOC means to display 3-level headers or 4-level headers, since the doc page may not contain a reference table, I think this design is quite common in markdown static websites, you can also check some jekyll themes as a reference.

**Reply by VarLad** [reply]
@Roger-luo Pluto generates a bunch of `pluto-cell`s I think 

Ever tried copying pasting the pluto cells from inside the container to the side...?
For this, someone needs to be more intimate with both Pluto and Franklin... Also, there's the problem of colliding headers (the styles techically). I don't think normally one would occur though. 
What I do know from this is, if we do implement a solution like that, the users themselves have to be intimate enough with HTML to solve any issues and the implementation won't be very general.

Solutions involving iframes basically display a complete site within a site. Displaying two different web pages in one. Thats why there's problems with its implementations I guess. 
But I think iframes solution would be much easier to use if the users get problems (because then they only have to worry about how to fix the iframe as one entity)

**Comment by VarLad** [comment]
@Roger-luo I ran into some problems with my iframes implementation today (with the docs integration)

I can fix that today, but I guess we'll get more issues like this in the future.
We need a solution using `divs`
For now I can try writing a CI which somehow weaves the header and the docs (based on your patch)
As for your idea of allowing more general HTML manipulation in Documenter, I think I'd need better instructions on what to make. 
~~As for the 2nd Documenter issue, I can make a function to parse the md files and make a div of ToC out of them. Its easy enough for the maintainer to apply that function in Documenter, so if he ever wants this feature, it'll take him 5-10 mins to integrate it into Documenter~~
Edit: Instructions just came from the maintainer. _On it!_

**Reply by VarLad** [reply]
Huhhh... any ideas where to go with the header/navbar-in-docs?

**Reply by VarLad** [reply]
@Roger-luo For the final problem: https://github.com/tlienart/Franklin.jl/issues/474#issuecomment-856085067 :)

This sounds interesting (although time taking). Basically a package with a lot of hfuns users can import!

**Reply by Roger-luo** [reply]
I see, that means some sort of plugin system needs to be built for Franklin then

**Reply by VarLad** [reply]
Yep! A package which has some custom hfuns which others can import!
This needs a better look into how Documenter works (specifically the fucntions that generate the HTML)!

**Comment by VarLad** [comment]
@Roger-luo Regarding https://github.com/JuliaDocs/Documenter.jl/issues/1563

Just FYI, I _can inject_ this modification into the generated docs using a CI, _without_ adding it as a feature to Documenter.jl. Seems easy enough.

**Reply by Roger-luo** [reply]
it seems Documenter people prefer us experimenting this outside the package first, so yeah let's do the evil thing to change the modification first. I assume mostly this only requires a CSS change and an extra TOC generation. The TOC can be generated from the Markdown AST directly without implementing your own parser.

**Comment by VarLad** [comment]
![Screenshot from 2021-06-06 20-02-58](https://user-images.githubusercontent.com/51269425/121199732-cb80e500-c890-11eb-8124-a96a76e22795.png)

I just used https://tscanlin.github.io/tocbot/

**Reply by Roger-luo** [reply]
this looks good to me in general, can we open a PR for this first?

**Reply by VarLad** [reply]
@Roger-luo https://github.com/QuantumBFS/Yao.jl/pull/284

Demo: https://varlad.github.io/Yao.jl/dev/man/registers.html

**Reply by VarLad** [reply]
The code is outright demonic but kinda works! :)
